### PR TITLE
AZD can use containers instead of code for deployment

### LINKED ISSUE #339: Dual deployment strategy. (azd and containers)
# Motivation

"Chat with your data" Solution accelerator currently lacks a "one-click deployment" option in the main branch, which is a critical feature for streamlined setup and user experience. 

# How would you feel if this feature request was implemented?

convo

# Requirements

The one-click deployment is a significant ease-of-use feature that can help accelerate adoption and improve the initial experience for new users of the "Chat with your data" Solution accelerator. By simplifying the deployment process, we can lower the barriers to entry and help users focus on leveraging the solution's capabilities rather than being bogged down by setup procedures.

# Tasks

_To be filled in by the engineer picking up the issue_

- [ ] Reintroduce the build and push of the 3 containers on a build of main
- [ ] Configure the App Service host bicep files to be configurable whether they deploy code or container.

- Ensuring that the ARM templates and deployment scripts in the ross/acr branch are up-to-date and compatible with the main branch.
- Updating the main branch documentation to include the one-click deployment instructions.
- Testing the merged feature thoroughly to ensure it works as expected in the main branch environment.

### PR DESCRIPTION
**Pull request by ross-p-smith**
## Purpose
This work enables an environment variable `AZURE_APP_SERVICE_HOSTING_MODEL` to be set which can be `code` or `container`. By default it will be container, which will pull the 3 packages Admin, Frontend, or Backend from a public Container Registry. This will greatly speed up the deployment as the code will already be built and App Service will just need to pull the container.

This has been tested by deploying a fresh instance, ingesting a document and then asking a question.

image

Fixes #339 

This fix will introduce new App Service names, so I recommend people run `azd down` and then `azd up` in order to get the cleanest build.

I will also add the "Deploy to Azure" back in as part of #378 when this hits main

## Does this introduce a breaking change?

```
[x] Yes
[] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

### CODE REVIEW

File: infra/main.bicep [hunk 1]
```diff
@@ -158,6 +158,12 @@ param principalId string = ''
 ])
 param authType string
 
+@allowed([
+  'code'
+  'container'
+])
+param hostingModel string
```
Comment (cecheta) [inline infra/main.bicep:165]: Could this default to `container`?

File: infra/main.bicepparam [hunk 2]
```diff
@@ -10,3 +10,5 @@ param principalId = readEnvironmentVariable('AZURE_PRINCIPAL_ID', 'principal_id'
 param useKeyVault = bool(readEnvironmentVariable('USE_KEY_VAULT', 'true'))
 
 param authType = readEnvironmentVariable('AZURE_AUTH_TYPE', 'keys')
+
+param hostingModel = readEnvironmentVariable('AZURE_APP_SERVICE_HOSTING_MODEL', 'container')
```
Comment (cecheta) [inline infra/main.bicepparam:14]: Is this value what is used when someone deploys from the command line? If so, should it instead default to `code`?

File: .github/workflows/build-docker-admin.yml [hunk 3]
```diff
@@ -0,0 +1,28 @@
+name: Admin Web App Docker Image
```
Comment (cecheta) [inline .github/workflows/build-docker-admin.yml:1]: Maybe we should only run these after `CI` finishes successfully?

File: Makefile [hunk 4]
```diff
@@ -44,3 +44,9 @@ deploy: azd-login ## 🚀 Deploy everything to Azure
 	@azd deploy web --no-prompt
 	@azd deploy function --no-prompt
 	@azd deploy adminweb --no-prompt
+
+docker-build: ## 🐳 Build the Docker image
```
Comment (cecheta) [inline Makefile:48]: What's the need for this? We're not calling it in the workflow, but the fact that it's hardcoded to `fruoccopublic.azurecr.io` seems like we wouldn't want to run it locally either?

File: infra/main.bicep [hunk 5]
```diff
@@ -11,7 +11,7 @@ param resourceToken string = toLower(uniqueString(subscription().id, environment
 param location string
 
 @description('Name of App Service plan')
-param hostingPlanName string = '${environmentName}-hosting-plan-${resourceToken}'
+param hostingPlanName string = 'hosting-plan-${resourceToken}'
```
Comment (cecheta) [inline infra/main.bicep:14]: How does this renaming affect resources that are already deployed?

**Comment by superhindupur** [comment]
Reviewed this with Ross over a teams call. Here is a summary of what we discussed:

- The three dockerfiles are added for admin, frontend and backend services, since before this change, the deployment was code-based only.
- The main.bicep file and related bicep files contain changes that introduce the parameter needed for deciding between code/container deployments. There is also a bugfix in there that shortens the names of the generated app services (unrelated to this change)
- role.bicep was removed since it contained duplicate code/config. 

Testing:

- Ross tested this by running the github actions from his branch and the app was deployed successfully using the containers in the registry. (Containers were pushed to the registry from his branch too, which was essentially the same as main except the infra configuration).

Based on the above, I am happy to approve this PR.

**Review by superhindupur — DISMISSED** [review]

File: Makefile (hunk 4 above)
Comment (ross-p-smith) [reply Makefile]: I've removed these as was just helpful whilst debugging - but we don't call them and you are right nobody knows the key to push them to that registry

File: infra/main.bicep (hunk 5 above)
Comment (ross-p-smith) [reply infra/main.bicep:14]: I believe that bicep does not manage state of changed resources, so I think that we are going to get newly create resources in peoples resource groups. I deially we need a changelog to say that breaking chnages are coming - what do you think?

File: infra/main.bicep (hunk 1 above)
Comment (ross-p-smith) [reply infra/main.bicep:165]: I'm following the pattern that is used for `authType` above. The bicepparam file will pass in a default of container.

File: infra/main.bicepparam (hunk 2 above)
Comment (ross-p-smith) [reply infra/main.bicepparam:14]: I've added an extra step in the README. I want the people who dont read the docs to ultimately click "Deploy to Azure" use containers and those that tinker with code to explicitly change it. In the future we could actually supply a Container Registry as part of the build - but trying to keep the scope down.

File: infra/main.bicep (hunk 5 above)
Comment (ross-p-smith) [reply infra/main.bicep]: @cecheta I've pushed and only changed the app services

File: Makefile [hunk 6]
```diff
@@ -45,8 +45,6 @@ deploy: azd-login ## 🚀 Deploy everything to Azure
 	@azd deploy function --no-prompt
 	@azd deploy adminweb --no-prompt
 
-docker-build: ## 🐳 Build the Docker image
```
Comment (superhindupur) [inline Makefile]: Don't we need this anymore?

Comment (ross-p-smith) [reply Makefile]: Github Actions carries out the build rather than locally - this was me testing locally. We don't actually "need" it so removed it

**Review by superhindupur — APPROVED** [review]

File: infra/main.bicep (hunk 5 above)
Comment (cecheta) [reply infra/main.bicep]: It looks like a number of them are still changing?

image

**Review by cecheta — APPROVED** [review]
Nice one :)

**Pull request merged**